### ISSUE TITLE
Module Federation: unable to share externalized dependency between host and remote

### ISSUE THREAD

**Issue by lubieowoce** [webpack/webpack#15554]
# Bug report

**TLDR**: `ModuleFederationPlugin` won't expose a host's dependency to a remote if the dependency is an `external`.  
To be honest, I'm not completely sure if this is a bug, a missing feature, or just a technical limitation of `ModuleFederationPlugin`... but i'd like to be able to do that 😄

> **NOTE**: i'm using React as an example, but this behavior occurs regardless of what library i'm trying to externalize+share.

---

### Context

Our module federation setup is pretty standard (except for the externals, apparently). We've got two apps:

```
(Host)           (Remote)
SomePage  ---->  SomeWidget
```
Both of these are React apps, with React stuff shared via a typical `ModuleFederationPlugin` config:
```
shared: {
  'react': { singleton: true },
  'react-dom': { singleton: true }
}
```
This all works fine.

Now, we'd like `SomePage` to get React from a CDN link, so we tried externalizing it in `SomePage` via a typical
```
externals: {
  'react': 'SomeGlobalVar',
  'react-dom': 'AnotherGlobalVar',
}
```
the idea being that React would get wired up everywhere like this:
```
Script tag  <script src="/our-cdn/react-stuff.js>
      |
      |
      |   webpack `external` (global var)
      |
      v
SomePage (Host app)
      |
      |
      |    ModuleFederationPlugin's `shared` exposes React
      |
      v
SomeWidget (Remote component)
```

### What is the current behavior?
Unfortunately, despite `SomePage` successfully loading React from the CDN, this doesn't work.
**The externalized React doesn't end up getting get shared between `SomePage` and `SomeWidget`**, so we get two copies of React in one tree, which leads to a crash for the usual reasons. From my testing, it appears that despite the `ModuleFederationPlugin` config, `SomePage` never puts React into the shared WMF scope, so `SomeWidget` understandably loads its fallback. The versions are the same between both, and removing them doesn't change anything. I've also tried this with lodash, and that also results in two copies, one from the CDN, and one loaded via `SomeWidget`'s fallback.

### What is the expected behavior?
Dependencies should be shared between host and remote even if they're externals, as illustrated in the diagram a few paragraphs up.

### Steps to reproduce
I've added some externals to `basic-host-remote` from `module-federation-examples` to reproduce this setup:

https://github.com/lubieowoce/module-federation-examples/tree/externals-test/basic-host-remote

the change is [very small](https://github.com/lubieowoce/module-federation-examples/commit/27f38561f87eda727d9c4d510bad8b4977af1cf5). You can run it via `cd basic-host-remote && yarn && yarn start`, then navigating to http://localhost:3001/. It'll flash and then turn blank, because when you when mix two copies of React in the same tree, it'll crash with "Error: Invalid hook call". This demonstrates that the library isn't actually being shared between the host and remote.

### Other relevant information:
webpack version:  `5.70.0`
Node.js version: `14.17.6`
Operating System: macOS 11.6, ARM

**Comment by vankop** [comment]
I think you should use such config
```
externals: {
  'react': 'global var',
  'react-dom': 'global var',
}
```
in both apps, `external` kind of mean to be not part of "build", so sharing it sounds weird.

**Comment by vankop** [comment]
I'm not sure that every external module could be shared... ( there is a lot of external module types )

**Comment by lubieowoce** [comment]
yeah, i imagine it's not easy, esp. since `ModuleFederationPlugin` itself seems to add an external for the remote... But do you think there's a chance to leave this open as a feature request? Or is this just a `WONTFIX`

**Comment by ScriptedAlchemy** [comment]
So the problem here. I think, is the remote container initialization phase. If a host marks it as external the runtime should still register react, and just point to the global variable. 
So the share scope still contains the keys to pass them into remote containers. Otherwise if my remote has import:false set - it crashes because share scope is missing keys, but loadSingleton is still part of the logic

**Comment by lubieowoce** [comment]
@ScriptedAlchemy tbh it's been a while since i've had this problem, but yeah, iirc being able to just stick the externalized react in the share-scope (if i remember my terminology right) would do the job, i.e. the remote module would then be able to pick it up. but i didn't have enough experience with webpack externals to do that myself 😅

also not sure what to do with the version-negotiation stuff -- how do we deal with the fact that we don't (and can't) know what version the externalized module has? make the user specify "i pinky promise it'll be react@17.0.2"? or perhaps fallback it to some magic `*`/ `ALWAYS_WINS_NEGOTIATION` value, since the user clearly doesn't want the module loaded via webpack? or maybe allow choosing either, with a config switch...? hairy stuff...

**Comment by ScriptedAlchemy** [comment]
Well if the webpack still registered externals in its api/share scope. Then you could manually specify the version and make the import value be external React 

That way share scope registers the global variable along with its version and if it's singleton or not. 

Ideally it's still tracked, just not imported. 

If you had mixed remotes. Where some may not set external, things would break. So this would be nice to still just keep the register() methods in the runtime. 

This causes me problems in node envs where react isn't bundled. Just require()d - I had to make sure all remotes treated react as external too because the host is missing keys.

**Comment by lubieowoce** [comment]
> I had to make sure all remotes treated react as external too because the host is missing keys.

Yeah, this is kind of what we did, except we made the remote in question have two builds - one w/ externalized react (for interop with stuff that externalizes react), and one "vanilla" (for WMF-native things where sharing Just Works). annoying but it works i guess

anyway, i think we agree that it's a useful feature, so do you think this issue should be reopened? like, for me, i no longer have a pressing need for this, so i'm fine either way. but maybe there's someone out there who needs it

(also if you know a way to make sharing externals happen "in userspace", i'd love to see how! mostly out of curiosity but also hey, might come in handy in the future)

**Comment by ScriptedAlchemy** [comment]
I think it should be reopened, this can cause strange regressions where there doesn't need to be. 
I believe what would need to change is the SharingRuntimeModule.js 

Regarding sharing externals in userspace; something like this might work. 

```js
{
  shared: {
    fakeReactKey: {
      import: 'var React',
      import: 'external React',
      import: 'data:text/javascript,module.exports = window.React;',
      shareKey: 'react'
    }
  }
}
```

But we would need a way to ensure it isnt tree shaken out as being "unused" in the graph.

**Comment by ScriptedAlchemy** [comment]
@vankop @sokra - any ideas on how we can make this happen?

**Comment by alexander-akait** [comment]
@ScriptedAlchemy Do we need reopen this?

**Comment by ScriptedAlchemy** [comment]
yes please!

**Comment by flyinhigh** [comment]
same problems, I find remoteEntry will build externals module and ignore share modules, if the are the same. but the host not use externals , it will throw a error , TypeError: Cannot read properties of undefined (reading 'React')

**Comment by ScriptedAlchemy** [comment]
That's because share scope is omnidirectional and the initialization phase is called by all runtimes who then register what they can offer, none of it loads at that point but they need to be callable since hosts call initExternal() which is remote.init()

**Comment by bh3605** [comment]
If you guys figured this out then you could have an MFE specify an external and the root could specify this same dependency as both external and a module federated shared dependency?

**Comment by bh3605** [comment]
@alexander-akait  Reopen again?

**Comment by ScriptedAlchemy** [comment]
Here is a solution to it, using runtime plugins:

https://github.com/module-federation/module-federation-examples/blob/master/federated-npm/app1/npm-runtime-global-plugin.js

https://github.com/module-federation/module-federation-examples/blob/master/federated-npm/app1/webpack.config.js

**Comment by skumar7-godaddy** [comment]
Hi,

I am trying to consume a remote  module built with React, React DOM and React-Router as shared scope inside a NextJS app which has a dependency that externalises React and ReactDOM.

I am getting an error when the code is trying to access React
<img width="1728" alt="Screenshot 2024-05-08 at 9 38 18 AM" src="https://github.com/webpack/webpack/assets/111855094/905afc59-7e62-431f-8df4-357985edb549">

The React and ReactDOM libraries are attached on the window object inside a ux property (window.ux.React and window.ux.ReactDOM)

This is my next.config webpack setting

`webpack: (config, options) => {
      const { isServer } = options;

      if (!isServer) {
        const federationConfig = {
          name: 'auth-ui',
          filename: 'static/chunks/remoteEntry.js',
          remotes: {
            // multi_users: 'multi_users@http://localhost:9000/remoteEntry.js',
            multi_users: 'multi_users@https://img6.wsimg.com/auth-assets/multi-users/remoteEntry.js',
          },
          // runtimePlugins: [require.resolve('./npm-runtime-global-plugin')],
          shared: {
            react: {
              eager: true,
              singleton: true,
            },
            'react-dom': {
              eager: true,
              singleton: true,
            },
            'react-router-dom': {
              eager: true,
              singleton: true,
            },
          },
        };
        config.plugins.push(
          new NextFederationPlugin(federationConfig),
        );
      }
      return config;
    },
  }`

Could someone please help ?

**Comment by bh3605** [comment]
> Here is a solution to it, using runtime plugins:
> 
> https://github.com/module-federation/module-federation-examples/blob/master/federated-npm/app1/npm-runtime-global-plugin.js
> 
> https://github.com/module-federation/module-federation-examples/blob/master/federated-npm/app1/webpack.config.js

@ScriptedAlchemy runtimePlugin isn't a property you can specify for the ModuleFederationPlugin class. What is that suppose to be?

**Comment by weilephp** [comment]
> global var
hi, I had same problem and I external in both host and remote, but not work, how do you work?

**Comment by diegomarcelino92** [comment]
Hi, I have a production case. My team uses Single-SPA to manage MFEs, other teams use WMF. In single-spa react is managed by system js and imports maps, so we want to share the react library loaded from SystemJS, is it possible? I have tried this code without sucsess:

```js
// single-spa MFE (using as host for other MFE with WBF
  const version = '18.0.2'

  await __webpack_init_sharing__('default')

  __webpack_share_scopes__.default.react = {
    [version]: {
      loaded: 1,
      from: '@org',
      get: System.import('react'),
    },
  }

  __webpack_share_scopes__.default['react-dom'] = {
    [version]: {
      loaded: 1,
      from: '@org',
      get: System.import('react-dom'),
    },
  }
  
  
  // Single spa MFE (host)
  new container.ModuleFederationPlugin({
        name: '@org/project,
        remotes: {
          '@project2': 'host@http://localhost:3000/remoteEntry.js',
        },
      }),
      
  // MFE federated
  new container.ModuleFederationPlugin({
    name: 'host',
    filename: 'remoteEntry.js',
    remotes,
    exposes: {
      './appfull': './src/ui/AppRoot.tsx',
    },
    shared: ['react', 'react-dom'],    
    })
```

**Comment by bh3605** [comment]
@diegomarcelino92 You'll need more libraries to get this working. `@module-federation/runtime` and `@module-federation/enhanced/webpack`. This is how I'm able to use a module federated app check a SystemJs mapping for shared dependencies.

```
/**
 * SystemJsRuntimePlugin for Module Federation
 * --------------------------------------
 * This plugin extends the capabilities of the module federation runtime by modifying the resolution of shared modules.
 *
 * Functionality:
 * 1. The plugin hooks into the 'resolveShare' hook of the federation runtime.
 * 2. It updates the references to shared module factories, allowing them to point to external sources like unpkg or esm.sh.
 * 3. This is particularly useful for cases where certain packages (like 'lodash') are preferred to be loaded from these external sources rather than bundled.
 *
 * How it Works:
 * - The plugin defines a 'useLocalShares' set specifying which modules should use local resolution.
 * - The 'getShareFromUnpkg' function dynamically imports a module from esm.sh, given a package name and version.
 * - In the 'resolveShare' hook, the plugin checks if the package should use local resolution (based on 'useLocalShares').
 * - If not, it updates the package reference to fetch from the external source.
 * - The 'beforeLoadShare' hook contains a workaround for an old issue which may not be relevant anymore.
 *
 * Usage:
 * - Register the plugin using 'registerGlobalPlugins' to enable its functionality in the module federation context.
 *
 * Notes:
 * - This plugin provides a flexible way to control module resolution, optimizing bundle sizes and leveraging CDN-hosted modules when desirable.
 */

 import { registerGlobalPlugins } from '@module-federation/runtime';

 const useLocalShares = new Set(['tslib']);

 const getShareFromSystemJs = (packageName, version) => {
   return () => {
     const mod = new Function(
       'packageName',
       'version',
       'return System.import(`${packageName}`)'
    //    'return import(/* webpackIgnore: true */ `https://esm.sh/${packageName}@${version}`)',
     )(packageName, version);
     return mod.then(m => {
       return () => m;
     });
   };
 };

 const store = {};

 const SystemJsRuntimeGlobalPlugin =
//   export default
   () => {
   return {
     name: 'share-from-systemjs-plugin',
     beforeInit: args => {
       store.name = args.options.name;
       return args;
     },

     resolveShare: args => {
       const { shareScopeMap, scope, pkgName, version, resolver } = args;
       const currentPackageRef = shareScopeMap[scope][pkgName][version];

       args.resolver = () => {
         if(!useLocalShares.has(pkgName)) {
           currentPackageRef.get = getShareFromSystemJs(pkgName, version);
         }
         return resolver();
       };

       return args;
     },
     beforeLoadShare: async args => {
       // old workaround, may not be required anymore # NOT NEEDED IN 0.3.1
    //    while (__FEDERATION__.__INSTANCES__.length <= 1) {
    //      await new Promise(r => setTimeout(r, 50));
    //    }
       return args;
     },
   };
 };

 registerGlobalPlugins([SystemJsRuntimeGlobalPlugin()]);

 export default () => ({
   name: 'empty-plugin',
 });
```

```
const { shareAll, share } = require('@angular-architects/module-federation/webpack');
const { ModuleFederationPlugin } = require('@module-federation/enhanced/webpack');
// const singleSpaAngularWebpack = require('single-spa-angular/lib/webpack').default;
const deps = require('./package.json').dependencies;

const modfedOptions = {
    library: { type: "module" },
    filename: "remoteEntry.js",
    name: 'MyApp',
    exposes: {
        './App': './src/app/app-remote-bootstrap.ts'
    },
    shared: //{ ...deps },
    {
        '@angular/core': {
            singleton: true,
            strictVersion: false
        },
        '@angular/core/rxjs-interop': {
            singleton: true,
            strictVersion: false
        },
        "@angular/common": {
            singleton: true,
            strictVersion: true,
            requiredVersion: deps["@angular/common"],
        },
        "@angular/common/http": {
          singleton: true,
          strictVersion: true,
          requiredVersion: deps["@angular/common/http"],
        },
        "@angular/router": {
            singleton: true,
            strictVersion: true,
            requiredVersion: deps["@angular/router"],
        },
        "@angular/platform-browser": {
            singleton: true,
            strictVersion: true,
            requiredVersion: deps["@angular/platform-browser"],
        },
        "@angular/elements": {
            singleton: true,
            strictVersion: true,
            requiredVersion: deps["@angular/elements"],
        },
        '@angular/forms': {
            singleton: true,
            strictVersion: true,
            requiredVersion: deps["@angular/forms"],
        },
        "@upmc-healthplan-npm/oidc-client": {
            singleton: true,
            strictVersion: true,
            requiredVersion: deps["@upmc-healthplan-npm/oidc-client"],
        },
        "rxjs": {
            singleton: true,
            strictVersion: true,
            requiredVersion: deps["rxjs"],
        },
    },
    // share({
    //     '@angular/core': {
    //         singleton: true,
    //         strictVersion: true,
    //         requiredVersion: 'auto',
    //       }
    //     // {
    //     //     ...shareAll({ singleton: true, strictVersion: true, requiredVersion: 'auto' }),
    //     // },
    // }),
    runtimePlugins: [require.resolve('./systemjs-runtime-global-plugin')],
    dts: false,
    disableManifest: true
};

module.exports = (config, options) => {
//   const singleSpaWebpackConfig = singleSpaAngularWebpack(config, options);

    config.experiments = {
        outputModule: true,
        topLevelAwait: true
    };
    config.output = {
      ...config.output,
      uniqueName: "MyApp",
      // depending on what imports the remote entry determines if this succeeds or not
      publicPath: "auto",
    };
    config.optimization = {
        ...config.optimization,
        runtimeChunk: false
    };
    config.plugins = [
        ...config.plugins,
        new ModuleFederationPlugin(modfedOptions),
    ];

  return config;
};
```

I use Angular and not React. However, you'll be able to use this with some tweaking.

**Comment by diegomarcelino92** [comment]
Wow, thank you very much @bh3605, it works well, so I simplified the code and the @module-federation/runtime package is not needed just @module-federation/enhanced/works, let me show you the code:

```js
// wmf.plugin.js
const useLocalShares = new Set(['react', 'react-dom']);

const getShareFromSystemJs = (packageName, version) => {
  console.log(packageName, version);
  return loadModule(packageName);
};

const loadModule = (packageName) => async () => {
  const module = await window.System.import(packageName);
  return () => module;
};

export default function () {
  return {
    name: 'share-from-systemjs-plugin',
    beforeInit: (args) => {
      console.log(args);
      return args;
    },

    resolveShare: (args) => {
      const { shareScopeMap, scope, pkgName, version, resolver } = args;
      const currentPackageRef = shareScopeMap[scope][pkgName][version];
      console.log(args, 'hass', useLocalShares.has(pkgName));

      args.resolver = () => {
        if (useLocalShares.has(pkgName) && window.System) {
          currentPackageRef.get = getShareFromSystemJs(pkgName, version);
        }
        return resolver();
      };

      return args;
    },
    
    beforeLoadShare: (args) => {
      console.log(args);
      return args;
    },
  };
}

// webpack.config.js 

const { ModuleFederationPlugin } = require('@module-federation/enhanced/webpack')

...
new ModuleFederationPlugin({
  name: 'appName',
  filename: 'remoteEntry.js',
  remotes,
  exposes: {
    './appfull': './src/ui/AppRoot.tsx',
  },
  shared: {
    react: {
      singleton: true
    },
    'react-dom': {
      singleton: true
    },
  },
  runtimePlugins: [require.resolve('./wmf.plugin.js')],
})
```

**Comment by bh3605** [comment]
@diegomarcelino92 Thanks for improving the code! My code originated from enhanced webpack's examples. I couldn't figure out  how to get the export default working without the empty plugin that was already there. Clearly, you did. You can also see the original function was using esm's cdn. If someone figured out a plugin that could take in an array of plugins you could have dependencies in a systemjs mapping or any number of other cdns to resolve to before it ultimately uses the module federated one.

**Comment by ScriptedAlchemy** [comment]
Yeah I can. I'll open. A git issue to track. I'm currently having to patch issues in how webpack handles chunks and I've also now changed how entry startup runtime time is rendered. So getting these solved in webpack core would be better. Then I can sync it to rspack.

**Comment by alexander-akait** [comment]
@ScriptedAlchemy Yeah, maybe we can implement additional hooks to provide more free

**Comment by ScriptedAlchemy** [comment]
I have created git issue to track the sub issues ill open. 

https://github.com/webpack/webpack/issues/18809

I have already opened PR for container reference hoisting
